### PR TITLE
Implement Storage Mass KIP and apply to TN11

### PR DESCRIPTION
**Pull request by michaelsutton**
Implements the still unpublished storage mass KIP and applies it on the mempool level (for all nets) and to consensus level for TN11

### CODE REVIEW

File: mining/src/mempool/check_transaction_standard.rs [hunk 1]
```diff
@@ -191,7 +199,8 @@ impl Mempool {
                 }
             }
 
-            let minimum_fee = self.minimum_required_transaction_relay_fee(transaction.calculated_mass.unwrap());
+            // For now, until wallets adapt, we don't require fee as function of full contextual_mass (but the fee/mass ratio will effect tx selection to block template)
```
Comment (someone235) [inline mining/src/mempool/check_transaction_standard.rs:202]: effect -> affect

File: mining/src/mempool/model/tx.rs [hunk 2]
```diff
@@ -22,7 +22,7 @@ impl MempoolTransaction {
     }
 
     pub(crate) fn fee_rate(&self) -> f64 {
-        self.mtx.calculated_fee.unwrap() as f64 / self.mtx.calculated_mass.unwrap() as f64
+        self.mtx.calculated_fee.unwrap() as f64 / self.mtx.calculated_compute_mass.unwrap() as f64
```
Comment (someone235) [inline mining/src/mempool/model/tx.rs:25]: I think it should use contextual mass

File: mining/src/mempool/validate_and_insert_transaction.rs [hunk 3]
```diff
@@ -68,7 +68,16 @@ impl Mempool {
...
+        // Calc the new mass including storage mass
+        let contextual_mass = consensus
+            .calculate_transaction_storage_mass(&transaction)
+            .and_then(|m| m.checked_add(transaction.calculated_compute_mass.unwrap()))
+            .ok_or(RuleError::RejectStorageMassIncomputable(transaction_id))?;
+
+        self.validate_transaction_in_context(&transaction, contextual_mass)?;
+
+        // Set the inner mass field post all validations
+        transaction.tx.set_mass(contextual_mass);
```
Comment (someone235) [inline mining/src/mempool/validate_and_insert_transaction.rs:80]: This field is used inside `validate_mempool_transaction` so it should be calculated there or before

File: mining/src/mempool/validate_and_insert_transaction.rs [hunk 4]
```diff
@@ -102,7 +111,8 @@ impl Mempool {
         Ok(())
     }
 
-    fn validate_transaction_in_context(&self, transaction: &MutableTransaction) -> RuleResult<()> {
+    fn validate_transaction_in_context(&self, transaction: &MutableTransaction, contextual_mass: u64) -> RuleResult<()> {
+        // TODO: seems like this can be removed now
```
Comment (someone235) [inline mining/src/mempool/validate_and_insert_transaction.rs:115]: So remove?

**Review by someone235 — CHANGES_REQUESTED** [review]

File: consensus/src/processes/transaction_validator/mod.rs [hunk 5]
```diff
@@ -33,6 +40,32 @@ impl TransactionValidator {
...
+            ghostdag_k,
+            coinbase_payload_script_public_key_max_len,
+            coinbase_maturity,
+            sig_cache: Cache::with_counters(10_000, counters),
+            mass_calculator,
+            storage_mass_activation_daa_score,
+        }
+    }
+
+    pub fn new_for_tests(
```
[the remark this answers is not in the document]
Comment (michaelsutton) [reply consensus/src/processes/transaction_validator/mod.rs:60]: feels more correct to create the mass calculator only once per consensus and then clone

File: consensus/src/pipeline/body_processor/body_validation_in_isolation.rs [hunk 6]
```diff
@@ -59,14 +59,29 @@ impl BlockBodyProcessor {
...
+                // This is only the compute part of the mass, the storage part cannot be computed here
+                let calculated_tx_compute_mass = self.mass_calculator.calc_tx_compute_mass(tx);
+                let committed_contextual_mass = tx.mass();
+                // We only check the lower-bound here, a precise check of the mass commitment
+                // is done when validating the tx in context
+                if committed_contextual_mass < calculated_tx_compute_mass {
+                    return Err(RuleError::MassFieldTooLow(tx.id(), committed_contextual_mass, calculated_tx_compute_mass));
+                }
+                // Sum over the committed masses
+                total_mass += committed_contextual_mass;
```
Comment (someone235) [inline consensus/src/pipeline/body_processor/body_validation_in_isolation.rs:73]: The miner can set anything in here, so we should check for overflows

File: consensus/src/pipeline/body_processor/body_validation_in_isolation.rs [hunk 7]
```diff
@@ -59,14 +59,29 @@ impl BlockBodyProcessor {
...
+                }
             }
-            total_mass += calculated_tx_mass;
-            if total_mass > self.max_block_mass {
-                return Err(RuleError::ExceedsMassLimit(self.max_block_mass));
+        } else {
+            for tx in block.transactions.iter() {
+                let calculated_tx_mass = self.mass_calculator.calc_tx_compute_mass(tx);
+                total_mass += calculated_tx_mass;
+                if total_mass > self.max_block_mass {
```
Comment (someone235) [inline consensus/src/pipeline/body_processor/body_validation_in_isolation.rs:82]: check overflow here too

**Review by someone235 — CHANGES_REQUESTED** [review]

**Review by someone235 — DISMISSED** [review]

**Review by someone235 — APPROVED** [review]

**Pull request merged**